### DISCUSSION TITLE
.NET MAUI updates in .NET 9 Release Candidate 1

### DISCUSSION THREAD

**Discussion by davidortinau** [dotnet/maui#24698] [Announcements]
Here's a summary of what's new in .NET MAUI in this preview release:

* [Added `HorizontalTextAlignment.Justify`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/rc1/dotnetmaui.md#added-horizontaltextalignmentjustify)

.NET MAUI updates in .NET 9 Release Candidate 1:
* [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/rc1/dotnetmaui.md) 
* [What's new in .NET MAUI in .NET 9](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9) documentation.
* [GitHub Release](https://github.com/dotnet/maui/releases/tag/9.0.0-rc.1.24453.9)

.NET 9 Preview RC1:
* [Discussion](https://aka.ms/dotnet/9/rc1)
* [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/rc1/README.md)

**Comment by jingliancui** [comment]
Hi team, when will release sr9 please? Looking forward to the window memory leak fix in sr9.

**Comment by DDHSchmidt** [comment]
The promised hot and new CollectionViewHandlers still didn't materialize, even in RC1:
![image](https://github.com/user-attachments/assets/3415811b-32da-4b8a-94e3-fce48e93ea83)

Visual Studio Version is Version `17.12.0 Preview 2.0`
dotnet--version is `9.0.100-rc.1.24452.12`

What am I missing @davidortinau ?

**Comment by baaaaif** [comment]
maybe the `<MauiVersion>9.xx</MauiVersion>` in your csproj?

**Reply by DDHSchmidt** [reply]
Thanks for your question @baaaaif , that forced me to take another look at our solution and we had an old Directory.build.props in our Solution directory that still imported an old version of Microsoft.Maui.Controls as part of a series of dozens of workarounds, necessary to circumvent the numerous issues in MAUI.

After that I was able to _really_ use .Net 9 and reference the new handlers. 👍

I'm still blocked in my upgrade tests, though.
- Apparently [Reference-Bindings are now considered evil](https://github.com/dotnet/maui/issues/23711), which baffles me. But I guess as long as the marketing department can now tout they're trimming complete, no one cares about the sacrifices? Anyway: Disabling compiled bindings for several bindings _could_ in theory help me with that.
- The bigger problem is: [CommunityToolkit.Maui plain out refuses to support .Net 9 until GA](https://github.com/CommunityToolkit/Maui/issues/2120#issuecomment-2289866990) and it's next to impossible to "just" rip it out of my project :/ This makes the point of testing out previews/ReleaseCandidates moot? 🤷

**Reply by Stedy59** [reply]
When a 3rd party library becomes a bottleneck, time to remove it.

**Reply by DDHSchmidt** [reply]
It depends... ripping out everything and writing it from scratch would qualify as the "Not invented here" syndrome 😅
Omitting the CommunityToolkit.Maui would feel like more work than it's worth right now, but we'll keep evaluating.

**Reply by ukehartmann** [reply]
@DDHSchmidt you're absolutely correct. Removing all the helpers just in order to test an RC is moot and rebuilding everything from scratch or copying the code over to your project is a silly proposal.

Did you also run into issues with enabling trim/AOT/r8? For us it trimmed all the DevExpress and CommunityToolkit libraries, which i had hoped would be fixed by .net9.

**Reply by DDHSchmidt** [reply]
> @DDHSchmidt you're absolutely correct. Removing all the helpers just in order to test an RC is moot and rebuilding everything from scratch or copying the code over to your project is a silly proposal.
> 
> Did you also run into issues with enabling trim/AOT/r8? For us it trimmed all the DevExpress and CommunityToolkit libraries, which i had hoped would be fixed by .net9.

Problems with AOT? Which flavour of it? We have the [oldschool AOT](https://learn.microsoft.com/en-us/dotnet/maui/macios/interpreter?view=net-maui-8.0#enable-the-interpreter) enabled by `<MtouchInterpreter>-all</MtouchInterpreter>` which has been around since the Xamarin days, I believe.
[It has](https://github.com/CommunityToolkit/Maui/issues/1752) [its edges](https://github.com/dotnet/maui/issues/13019), but generally it *is usable* and provides some performance benefits if you are prepared to thoroughly test your app and are prepared to dive deep... very deep into your logs and crash reports in case it breaks.

Then we have the official, [shiny new AOT](https://learn.microsoft.com/en-us/dotnet/maui/whats-new/dotnet-9?view=net-maui-8.0#native-aot-for-ios--mac-catalyst), available since .Net 8 and improved(?) in .Net 9.
My advice: Hands off that one.
There were glamorous blog posts written about it, touting great performacne benefits but the warning in the linked help article says it all: "Your app ***and it's dependencies*** must be fully trimmable in order to utilize this feature."

Translation: It's cool to use for toy apps, but unless all of your dependencies did their homework and got updated you shouldn't use this... at all.
Last time I experimented with it the crashes & errors were so cryptic that I couldn't find out where to start analyzing :/
I don't expect this to be usable before maybe .NET 11

**Reply by ukehartmann** [reply]
@DDHSchmidt the latter is what i've been experimenting with ever since .NET 7 and i can only concur. It's been useless and i had hopes that the glamorous blog posts would yield some results. So far it's been a letdown as the most helpful libraries we use do not get trimmed correctly and adding manual exceptions to the trim doesn't really make that option that desirable for quick improvements in file size and performance.

**Reply by Stedy59** [reply]
Point in case... If you use third party libraries, you live and die by that sword!

**Reply by ukehartmann** [reply]
Obviously, but some people prefer to ship products over not having any dependencies.

**Reply by Stedy59** [reply]
My statement wasn't "all encompassing." If you are having an issue with "one" of your dependencies, how are you shipping, let alone, in a state you desire or, require? Context AND perspective is important...

**Reply by ukehartmann** [reply]
@Stedy59 so you obviously do not use MAUI but build **everything** from scratch, right?

**Comment by atanubanerjee** [comment]
I've tried the new HybridWebView control in both preview 7 as well as RC1. I'm just trying out the example  in the online documentation (send messages from C# to JS and vice versa). This works fine on Windows, but on Android only the messages from C# to JavaScript work. Sending messages from JS to C# leads to an exception (System.Reflection.TargetInvocationException). I can provide a repro, but it's just the example in documentation. 

On a different app when I try this, I get a Java.Lang.RuntimeException Message=java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. 

Any suggestions? Are you already tracking this as a bug, or could the online documentation be wrong in some way? BTW I don't have this problem on .Net 8, using the experimental HybridWebView control on github from Eilon - as messages work fine in both directions for Android.
